src/server/frontend_xwayland/xwayland_surface.h

```diff
@@ -75,6 +91,52 @@ class XWaylandSurface
     void move_resize(uint32_t detail);
 
 private:
+    struct StateTracker
+    {
+        static auto make_withdrawn() -> StateTracker
+        {
+            return StateTracker{scene::SurfaceStateTracker(mir_window_state_restored), true};
+        }
+
+        auto operator==(StateTracker other) const -> bool
+        {
+            return mir_state == other.mir_state && withdrawn == other.withdrawn;
+        }
+
+        auto active_mir_state() const -> MirWindowState
+        {
+            return withdrawn ? mir_window_state_hidden : mir_state.active_state();
+        }
+        auto with_active_mir_state(MirWindowState state) const
+        {
+            return StateTracker{mir_state.with_active_state(state), false};
+        }
```

Comment:
I'm a bit confused by this design.

The idiom `auto with...(...) const -> NewObject` is usually associated with immutable types (and these have obvious benefits for thread safety), but `StateTracker` is not immutable: it supports copy and assignment.

So why not go one way or the other? Either:
1. Accept that `StateTracker` is mutable and have mutator methods; or,
2. Make `StateTracker` immutable and "own" the tracked instance through an owning pointer.

Reply:
`StateTracker`, much like `ms::SurfaceStateTracker`, is mutable only by assignment, similar to primitive value types. Making it mutable would lead to it being non-obvious where changes to it were being made. It would also make the class less ergonomic to use. Using a `unique_ptr<StateTracker>` would cost a bunch of allocations and pointer lookups and I don't see any benefit that would yield (unique pointer assignment is not threadsafe without additional locking, so it would leave us in the same place there). A `const StateTracker` can be used if assignment is not desired.